pkg/controller/certificate-shim/sync.go

```diff
@@ -275,7 +275,8 @@ func buildCertificates(
 	var newCrts []*cmapi.Certificate
 	var updateCrts []*cmapi.Certificate
 
-	tlsHosts := make(map[corev1.ObjectReference][]string)
+	tlsHostsPerSecretRef := make(map[corev1.ObjectReference][]string)
+	var secretRefs []corev1.ObjectReference // Work around map indeterminism.
```

Comment:
Suggestion: would it not make more sense to not use the 'unsorted' function for comparing events instead? Do we actually want to test that these two events are thrown in order? I can see that this would fix the issue, but it makes me a little uneasy that we are adding extra complexity here just so that the test passes if we don't actually care about the order of these events.
If we leave this in, could we perhaps expand on the explanation why this was done?

Reply:
You are right: just for the sake of making the results of `buildCertificates` deterministic, my approach makes things more complex (and potentially less performant, although I'd need to run benchmarks to prove it).

My thought process was that keeping indeterministic behaviors tends to "creep" into other parts of testing which makes me a bit uneasy... But it might not be that big of a deal in the end. 😅

@jakexks What do you think?

Reply:
It is slightly slower (twice as slow, to be precise):

```sh
cpu: AMD Ryzen 9 5950X 16-Core Processor
Benchmark/nondeterministic-32            1000000              1868 ns/op
Benchmark/deterministic-32               1331593               972.5 ns/op
```

<details>
<summary>bench code</summary>

```go
package main_test

import (
	"testing"
)

var letters = []string{"a", "b", "a", "c", "a", "d", "e", "e", "a", "b", "d", "f", "g", "h", "i", "j", "d", "i", "g", "e"}

func Benchmark(b *testing.B) {
	b.Run("nondeterministic", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			lettersCount := make(map[string]int)
			var lettersOrder []string

			for _, letter := range letters {
				if _, ok := lettersCount[letter]; !ok {
					lettersOrder = append(lettersOrder, letter)
				}
				lettersCount[letter]++
			}
		}
	})

	b.Run("deterministic", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			lettersCount := make(map[string]int)

			for _, letter := range letters {
				lettersCount[letter]++
			}
		}
	})
}

// Run with: go test -bench=.
```

</details>

I think it is worth the performance penalty since we avoid changing the order in which we expect events for other tests.